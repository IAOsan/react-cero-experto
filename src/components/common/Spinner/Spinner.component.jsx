import React from 'react';
import './spinner.styles.css';

function Spinner({ size = 'default', color = 'default', inline }) {
	const sizes = {
		xs: 'lds-ring lds-ring--xs',
		sm: 'lds-ring lds-ring--sm',
		default: 'lds-ring',
	};
	const colors = {
		light: '#f5f5f5',
		default: '#4d4c7d',
	};

	function renderSpinner() {
		const spinner = (
			<div
				style={{ '--c': colors[color] || colors['default'] }}
				className={sizes[size] || sizes['default']}
			>
				<div></div>
				<div></div>
				<div></div>
				<div></div>
			</div>
		);

		return inline ? (
			spinner
		) : (
			<div className='flex flex-jc-c'>{spinner}</div>
		);
	}

	return <>{renderSpinner()}</>;
}

export default Spinner;
