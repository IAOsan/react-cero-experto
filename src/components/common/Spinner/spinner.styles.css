:root {
	--c: #007feb;
}
.lds-ring {
	display: inline-block;
	position: relative;
	width: 8rem;
	height: 8rem;
}
.lds-ring--sm {
	width: 4rem;
	height: 4rem;
}
.lds-ring--xs {
	width: 2rem;
	height: 2rem;
}
.lds-ring div {
	box-sizing: border-box;
	display: block;
	position: absolute;
	width: 6.4rem;
	height: 6.4rem;
	margin: 0.8rem;
	border: 0.8rem solid var(--c);
	border-radius: 50%;
	animation: lds-ring 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
	border-color: var(--c) transparent transparent transparent;
}
.lds-ring--sm div {
	width: 3.2rem;
	height: 3.2rem;
	margin: 0.4rem;
	border: 0.4rem solid var(--c);
	border-color: var(--c) transparent transparent transparent;
}
.lds-ring--xs div {
	width: 1.6rem;
	height: 1.6rem;
	margin: 0.2rem;
	border: 0.2rem solid var(--c);
	border-color: var(--c) transparent transparent transparent;
}
.lds-ring div:nth-child(1) {
	animation-delay: -0.45s;
}
.lds-ring div:nth-child(2) {
	animation-delay: -0.3s;
}
.lds-ring div:nth-child(3) {
	animation-delay: -0.15s;
}
@keyframes lds-ring {
	0% {
		transform: rotate(0deg);
	}
	100% {
		transform: rotate(360deg);
	}
}
