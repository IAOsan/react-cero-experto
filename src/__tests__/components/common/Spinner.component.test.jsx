import 'vitest';
import '@testing-library/jest-dom';
import { render } from '../../test-utils';
import Spinner from '../../../components/common/Spinner/Spinner.component';

describe('<Spinner />', () => {
	it('should renders by default properly', () => {
		const wrapper = renderSpinner();
		expect(wrapper.container.firstChild).toMatchSnapshot();
	});

	it('should renders in size extra small', () => {
		renderSpinner({ size: 'xs' });

		const spinner = document.querySelector('.lds-ring');
		expect(spinner).toHaveClass('lds-ring--xs');
	});

	it('should renders in size extra small', () => {
		renderSpinner({ size: 'sm' });

		const spinner = document.querySelector('.lds-ring');
		expect(spinner).toHaveClass('lds-ring--sm');
	});

	it('should renders in color light', () => {
		renderSpinner({ color: 'light' });

		const spinner = document.querySelector('.lds-ring');
		expect(spinner).toHaveStyle('--c: #f5f5f5');
	});

	it('should not wrap into flex if its inline', () => {
		renderSpinner({ inline: true });

		expect(
			document.querySelector('.lds-ring').parentElement.className
		).toBeFalsy();
	});
});

function renderSpinner(props) {
	return render(<Spinner {...props} />);
}
