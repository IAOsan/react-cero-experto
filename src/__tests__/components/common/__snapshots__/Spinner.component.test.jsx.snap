// Vitest Snapshot v1

exports[`<Spinner /> > should renders by default properly 1`] = `
<div
  class="flex flex-jc-c"
>
  <div
    class="lds-ring"
    style="--c: #4d4c7d;"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;
